apis/apps/v1alpha1/containerrecreaterequest_types.go

```diff
@@ -0,0 +1,182 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// [Immutable] Pod name of this ContainerRecreateRequest.
+	ContainerRecreateRequestPodNameKey = "crr.apps.kruise.io/pod-name"
+	// [Immutable] Pod UID of this ContainerRecreateRequest.
+	ContainerRecreateRequestPodUIDKey = "crr.apps.kruise.io/pod-uid"
+	// [Immutable] Node name of this ContainerRecreateRequest.
+	ContainerRecreateRequestNodeNameKey = "crr.apps.kruise.io/node-name"
+
+	// [Mutable] ContainerRecreateRequestActiveKey indicates if this ContainerRecreateRequest is active.
+	// It will be removed in labels since a ContainerRecreateRequest has completed.
+	// We use it mainly to make kruise-daemon do not list watch those ContainerRecreateRequests that have completed.
+	ContainerRecreateRequestActiveKey = "crr.apps.kruise.io/active"
+	// [Mutable] ContainerRecreateRequestSyncContainerStatusesKey contains the container statuses in current Pod.
+	// It is only synchronized during a ContainerRecreateRequest in Recreating phase.
+	ContainerRecreateRequestSyncContainerStatusesKey = "crr.apps.kruise.io/sync-container-statuses"
+	// [Mutable] ContainerRecreateRequestReadinessClosedKey indicates the Pod has been forced to not-ready.
+	// It is required if the readyProtection is set in ContainerRecreateRequests.
+	ContainerRecreateRequestReadinessClosedKey = "crr.apps.kruise.io/readiness-closed"
+
+	// The minimum stopGracePeriod for containers to recreate.
+	ContainerRecreateRequestMinimumStopGracePeriodSeconds = 3
+)
+
+// ContainerRecreateRequestSpec defines the desired state of ContainerRecreateRequest
+type ContainerRecreateRequestSpec struct {
+	// Containers contains the containers that need to recreate in the Pod.
+	Containers []ContainerRecreateRequestContainer `json:"containers"`
+	// Strategy defines strategies for containers recreation.
+	Strategy *ContainerRecreateRequestStrategy `json:"strategy,omitempty"`
+	// ActiveDeadlineSeconds is the deadline duration of this ContainerRecreateRequest.
+	ActiveDeadlineSeconds *int64 `json:"activeDeadlineSeconds,omitempty"`
+	// TTLSecondsAfterFinished is the TTL duration after this ContainerRecreateRequest has completed.
+	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
+}
+
+// ContainerRecreateRequestContainer defines the container that need to recreate.
+type ContainerRecreateRequestContainer struct {
+	// Name of the container that need to recreate.
+	// It must be existing in the real Pod containers.
+	Name string `json:"name"`
+	// SkipPreStop indicates whether to skip the preStop hook during recreating this container.
+	SkipPreStop bool `json:"skipPreStop,omitempty"`
```

Comment:
is it necessary to include PreStop, ports and StatusContext fields? is it ok to  just use the config in pod spec